jitstats.lua

```diff
@@ -0,0 +1,462 @@
+
+function jitstats.setdebugprint(enable)
+    jitstats.debug = enable
+    debugprint = (enable and printf) or function() end
+end
+
+local function topercent(total, subval)
+    return math.floor(subval / total * 100)
+end
+
+function jitstats.print_simple(msg)
+    local running = jitstats.stop()
+    printf("%sStarted %d, Aborted %d, Exits %d", msg or "", stats.starts, stats.aborts, stats.exits)
+    if running then
+        jitstats.start()
+    end
+end
+
+local function table_keys(t)
+    local result = {}
+    for k,_ in pairs(t) do
+        table.insert(result, k)
+    end
+    return result
+end
+
+local function default_sorter(t, k1, k2)
+    return t[k1] > t[k2] 
+end
+
+local function table_sortkeys(t, sorter)
+    local keys = table_keys(t) 
+    sorter = sorter or default_sorter
+    table.sort(keys, function(k1, k2) return sorter(t, k1, k2) end)
+    return keys
+end
+
+local function statstbl_concat(tbl)
+    local strings = {}
+    local count = 0
+    for k, v in pairs(tbl) do
+        count = count + 1
+        strings[count] = k.." = "..v
+    end
+    
+    return table.concat(strings, ", ")
+end
+
+function jitstats.print(statstbl, msg)
+    local running = jitstats.stop()
+
+    if not statstbl then
+        statstbl = stats
+        msg = "Trace Stats:"
+    elseif type(statstbl) == "string" then
+        assert(msg == nil)
+        msg = statstbl
+        statstbl = stats
+    else
+        assert(type(statstbl) == "table", type(statstbl))
+        msg = "Trace Stats:"
+    end
+
+    printf(msg)
+    printf("  Started %d, side %d(%g%%), stitch %d", statstbl.starts, statstbl.side_starts, topercent(statstbl.starts, statstbl.side_starts), statstbl.stitch_starts)
+    printf("  Completed Link types: %s", statstbl_concat(statstbl.linktypes))
+    printf("  Aborted %d(%g%%), side %d, stitch %d", statstbl.aborts, topercent(statstbl.starts, statstbl.aborts), statstbl.side_aborts, statstbl.stitch_aborts)
+
+    local sortedkeys = table_sortkeys(statstbl.abort_counts)
+    
+    for _, name in ipairs(sortedkeys) do
+        local msg = (traceerr[name] and vmdef.traceerr[traceerr[name]]) or name
+        printf("    %d %s", statstbl.abort_counts[name], msg)
+    end
+
+    -- Print NYI bytecode abort counts
+    if next(statstbl.nyi_bc) then
+        printf("  NYI BC Aborts")
+        sortedkeys = table_sortkeys(statstbl.nyi_bc)
+        for _, bc in ipairs(sortedkeys) do
+            printf("    %d %s", statstbl.nyi_bc[bc], bc)
+        end
+    end
+
+    if statstbl.func_blacklisted > 0 then
+        printf(" Functions blacklisted: %d", statstbl.func_blacklisted)
+    end
+    if statstbl.loop_blacklisted > 0 then
+        printf(" Loops blacklisted: %d", statstbl.loop_blacklisted)
+    end
+    printf(" Uncompiled exits taken %d", statstbl.exits)
+
+    if running then
+        jitstats.start()
+    end
+end
+
+-- Turn off the jit for all the functions in this file so we don't contribute to the stats when being called. Sadly 
```

Comment:
This comment is a really long sentence and I can't quite parse the last part of it. Maybe break it up into two sentences?

Reply:
how about this @ltratt
> Turn off the jit for all the functions in this file so they don't try JIT. Sadly the effect
> of this call is deferred with respect to switching the function header bytecode. It is only set 
> to interpreter only mode the first time each function tries to start a trace.

Reply:
Let's capitalise "JIT" consistently. Unfortunately I still don't understand sentences 2 and 3 :(